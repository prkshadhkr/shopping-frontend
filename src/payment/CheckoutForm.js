import React, { useState, useEffect } from "react";
import {
  CardElement,
  useStripe,
  useElements
} from "@stripe/react-stripe-js";
import './Checkout.css';

import { BASE_URL } from '../constants/generic';
import { useSelector } from "react-redux";
import { TOKEN_IN_STORAGE } from '../App';
import { clear } from "redux-localstorage-simple";

export default function CheckoutForm() {
  const order = useSelector(st => st.orders.order);
  const order_id = Object(order).id;
  const _token = localStorage[TOKEN_IN_STORAGE];
  console.log('+++++++ order_id +++++++++', order_id);

  const [succeeded, setSucceeded] = useState(false);
  const [error, setError] = useState(null);
  const [processing, setProcessing] = useState('');
  const [disabled, setDisabled] = useState(true);
  const [clientSecret, setClientSecret] = useState('');
  const stripe = useStripe();
  const elements = useElements();
 


  useEffect(() => {
    // Create PaymentIntent as soon as the page loads
    window
      .fetch(`${BASE_URL}/orders/payment`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify({order_id, _token})
      })
      .then(res => {
        clear();
        return res.json();
      })
      .then(data => {
        setClientSecret(data.clientSecret);
      });
  }, []);

  const cardStyle = {
    style: {
      base: {
        color: "#32325d",
        fontFamily: 'Arial, sans-serif',
        fontSmoothing: "antialiased",
        fontSize: "16px",
        "::placeholder": {
          color: "#32325d"
        }
      },
      invalid: {
        color: "#fa755a",
        iconColor: "#fa755a"
      }
    }
  };

  const handleChange = async (event) => {
    // Listen for changes in the CardElement
    // and display any errors as the customer types their card details
    setDisabled(event.empty);
    setError(event.error ? event.error.message : "");
  };

  const handleSubmit = async ev => {
    ev.preventDefault();
    setProcessing(true);

    const payload = await stripe.confirmCardPayment(clientSecret, {
      payment_method: {
        card: elements.getElement(CardElement),
        billing_details: {
          name: ev.target.name.value
        }
      }
    });

    if (payload.error) {
      setError(`Payment failed ${payload.error.message}`);
      setProcessing(false);
    } else {
      setError(null);
      setProcessing(false);
      setSucceeded(true);
    }
  };

  return (
    <form id="payment-form" onSubmit={handleSubmit}>
      <CardElement id="card-element" options={cardStyle} onChange={handleChange} />
      <button className="payment-submit"
        disabled={processing || disabled || succeeded}
        id="submit"
      >
        <span id="button-text">
          {processing ? (
            <div className="spinner" id="spinner"></div>
          ) : (
            "Pay"
          )}
        </span>
      </button>
      {/* Show any error that happens when processing the payment */}
      {error && (
        <div className="card-error" role="alert">
          {error}
        </div>
      )}
      {/* Show a success message upon completion */}
      <p className={succeeded ? "result-message" : "result-message hidden"}>
        Payment succeeded, see the result in your
        <a
          href={`https://dashboard.stripe.com/test/payments`}
        >
          {" "}
          Stripe dashboard.
        </a> Refresh the page to pay again.
      </p>
    </form>
  );
}
