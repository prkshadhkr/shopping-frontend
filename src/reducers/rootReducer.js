import { combineReducers } from 'redux';
import products  from './products';
import reviews from './reviews';
import orders from './orders';

const rootReducer = combineReducers(
  { products, reviews, orders }
)

export default rootReducer;