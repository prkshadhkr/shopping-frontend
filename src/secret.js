const STRIPE_SECRET_KEY = "pk_test_51HDu1bKr580TeDUY4R0GSxXwNSYqqco64NW1yCeSkOrzIYRezdflvWZ3cZQeNVWOjyXKqRcGQpFca6E01ulVrS7f00mb3nw2f0"

export {
  STRIPE_SECRET_KEY
} 