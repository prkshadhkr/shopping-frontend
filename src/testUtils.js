import React from  'react';
import UserContext from './user/UserContext';

const demoUser = {
  username: "testing",
  first_name: "testing",
  last_name: "testing",
  email: "testing@gmail.com",
  password: "testing",
  is_admin: false
};

/**
 * user is destructured from userContext in Profilejs as
 * const { user, setUser } = useContext(UserContext);
 */
const UserProvider = ({ children, user = demoUser }) => (
  <UserContext.Provider value={{user}}>
    {children}
  </UserContext.Provider>
)

export { UserProvider };