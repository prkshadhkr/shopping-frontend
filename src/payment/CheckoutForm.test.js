import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { UserProvider } from '../testUtils';
import CheckoutForm from './CheckoutForm';
import rootReducer from '../reducers/rootReducer';

import { loadStripe } from "@stripe/stripe-js";
import { Elements } from "@stripe/react-stripe-js";
import { STRIPE_SECRET_KEY } from '../secret';

const promise = loadStripe(STRIPE_SECRET_KEY);
const store = createStore(rootReducer);


it ('should render without crashing', function() {
  render (
    <MemoryRouter>
      <Provider store={store}>
      <UserProvider>
        <Elements stripe={promise}>
        <CheckoutForm />
        </Elements>
      </UserProvider>
      </Provider>
    </MemoryRouter>
  )
});

it ('should match snapshot', function () {
  const {asFragment} = render (
    <MemoryRouter>
      <Provider store={store}>
        <UserProvider>
          <Elements stripe={promise}>
            <CheckoutForm />
          </Elements>
        </UserProvider>
      </Provider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
})

it ('should match snapshot if user logged out', function () {
  const {asFragment} = render (
    <MemoryRouter>
      <Provider store={store}>
        <UserProvider user={null}>
          <Elements stripe={null}>
            <CheckoutForm />
          </Elements>
        </UserProvider>
      </Provider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
})