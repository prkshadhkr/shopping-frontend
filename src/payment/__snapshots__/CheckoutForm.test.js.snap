// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="Payment d-flex justify-content-center "
  >
    <form
      id="payment-form"
    >
      <div
        id="card-element"
      />
      <button
        class="payment-submit"
        disabled=""
        id="submit"
      >
        <span
          id="button-text"
        >
          Pay
        </span>
      </button>
      <p
        class="result-message hidden"
      >
        Payment succeeded, see the result in your
        <a
          href="https://dashboard.stripe.com/test/payments"
        >
           Stripe dashboard.
        </a>
         Refresh the page to pay again.
      </p>
    </form>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot if user logged out 1`] = `
<DocumentFragment>
  <div
    class="Payment d-flex justify-content-center "
  >
    <form
      id="payment-form"
    >
      <div
        id="card-element"
      />
      <button
        class="payment-submit"
        disabled=""
        id="submit"
      >
        <span
          id="button-text"
        >
          Pay
        </span>
      </button>
      <p
        class="result-message hidden"
      >
        Payment succeeded, see the result in your
        <a
          href="https://dashboard.stripe.com/test/payments"
        >
           Stripe dashboard.
        </a>
         Refresh the page to pay again.
      </p>
    </form>
  </div>
</DocumentFragment>
`;
